File: src/hp_guard/adapter.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Callable, Mapping

from .audit import AuditError, AuditedPolicyStore, PolicySnapshot
from .models import Action, PolicyCall


_REQUEST_FIELDS = {"caller_id", "correlation_id", "deadline_unix_ms", "call"}
_CALL_FIELDS = {"agent", "tool", "args", "user", "context"}
_MAX_U64 = (1 << 64) - 1


class AdapterError(Exception):
    """An inline-adapter request or required authorization failure."""

    def __init__(self, code: str, message: str):
        self.code = code
        super().__init__(message)


@dataclass(frozen=True)
class EnforcementRequest:
    caller_id: str
    correlation_id: str
    deadline_unix_ms: int
    call: PolicyCall

    def __post_init__(self) -> None:
        _opaque_id(self.caller_id, "caller_id")
        _opaque_id(self.correlation_id, "correlation_id")
        _validate_deadline(self.deadline_unix_ms)
        object.__setattr__(self, "call", _normalize_policy_call(self.call))

    @classmethod
    def from_dict(cls, value: Any) -> EnforcementRequest:
        if not isinstance(value, Mapping) or set(value) != _REQUEST_FIELDS:
            raise AdapterError("invalid_request", "request must contain exactly the required fields")
        caller_id = _opaque_id(value["caller_id"], "caller_id")
        correlation_id = _opaque_id(value["correlation_id"], "correlation_id")
        deadline_unix_ms = value["deadline_unix_ms"]
        _validate_deadline(deadline_unix_ms)
        return cls(
            caller_id=caller_id,
            correlation_id=correlation_id,
            deadline_unix_ms=deadline_unix_ms,
            call=_normalize_call(value["call"]),
        )


@dataclass(frozen=True)
class EffectRequest:
    caller_id: str
    correlation_id: str
    deadline_unix_ms: int
    policy: PolicySnapshot
    decision: str
    matched_rules: tuple[int, ...]
    call: PolicyCall

    def to_dict(self) -> dict[str, Any]:
        return {
            "caller_id": self.caller_id,
            "correlation_id": self.correlation_id,
            "deadline_unix_ms": self.deadline_unix_ms,
            "policy": _policy_identity(self.policy),
            "decision": self.decision,
            "matched_rules": list(self.matched_rules),
            "call": _call_dict(self.call),
        }


@dataclass(frozen=True)
class EnforcementResponse:
    caller_id: str
    correlation_id: str
    deadline_unix_ms: int
    policy: PolicySnapshot
    decision: str
    matched_rules: tuple[int, ...]
    effect: EffectRequest | None = None

    def to_dict(self) -> dict[str, Any]:
        response = {
            "caller_id": self.caller_id,
            "correlation_id": self.correlation_id,
            "deadline_unix_ms": self.deadline_unix_ms,
            "policy": _policy_identity(self.policy),
            "decision": self.decision,
            "matched_rules": list(self.matched_rules),
        }
        if self.effect is not None:
            response["effect"] = self.effect.to_dict()
        return response


class InlineEnforcementAdapter:
    """Authorize a normalized call and return data for a host-owned effect."""

    def __init__(
        self,
        policy_store: AuditedPolicyStore,
        *,
        now_unix_ms: Callable[[], int] | None = None,
    ):
        self._policy_store = policy_store
        self._now_unix_ms = now_unix_ms or (lambda: time.time_ns() // 1_000_000)

    def authorize(self, request: EnforcementRequest) -> EnforcementResponse:
        if not isinstance(request, EnforcementRequest):
            raise AdapterError("invalid_request", "request must be an EnforcementRequest")
        call = _normalize_policy_call(request.call)
        if request.deadline_unix_ms <= self._now_unix_ms():
            raise AdapterError("deadline_exceeded", "request deadline has elapsed")
        try:
            authorization = self._policy_store._authorize_with_metadata(
                call,
                correlation_id=request.correlation_id,
                caller_id=request.caller_id,
                deadline_unix_ms=request.deadline_unix_ms,
            )
        except AuditError as error:
            raise AdapterError("audit_write_failed", "required audit write failed") from error

        decision = authorization.decision.action.value
        matched_rules = tuple(authorization.decision.matched_rules)
        effect = None
        if authorization.decision.action is Action.ALLOW:
            effect = EffectRequest(
                caller_id=request.caller_id,
                correlation_id=request.correlation_id,
                deadline_unix_ms=request.deadline_unix_ms,
                policy=authorization.snapshot,
                decision=decision,
                matched_rules=matched_rules,
                call=call,
            )
        return EnforcementResponse(
            caller_id=request.caller_id,
            correlation_id=request.correlation_id,
            deadline_unix_ms=request.deadline_unix_ms,
            policy=authorization.snapshot,
            decision=decision,
            matched_rules=matched_rules,
            effect=effect,
        )


def _opaque_id(value: Any, name: str) -> str:
    if not isinstance(value, str) or not value:
        raise AdapterError("invalid_request", f"{name} must be a non-empty string")
    size = _utf8_size(value, name)
    if size > 256:
        raise AdapterError("invalid_request", f"{name} must be at most 256 UTF-8 bytes")
    return value


def _validate_deadline(value: Any) -> None:
    if type(value) is not int or not 0 <= value <= _MAX_U64:
        raise AdapterError("invalid_request", "deadline_unix_ms must be an unsigned 64-bit integer")


def _utf8_size(value: str, name: str) -> int:
    try:
        return len(value.encode("utf-8"))
    except UnicodeEncodeError as error:
        raise AdapterError("invalid_request", f"{name} must be valid UTF-8") from error


def _normalize_call(value: Any) -> PolicyCall:
    if not isinstance(value, Mapping) or set(value) != _CALL_FIELDS:
        raise AdapterError("invalid_request", "call must contain exactly the required fields")
    for name in ("agent", "tool", "user"):
        if value[name] is not None and not isinstance(value[name], str):
            raise AdapterError("invalid_request", f"call.{name} must be a string or null")
        if value[name] is not None:
            _utf8_size(value[name], f"call.{name}")
    args = value["args"]
    if not isinstance(args, list) or any(not isinstance(argument, str) for argument in args):
        raise AdapterError("invalid_request", "call.args must be an array of strings")
    for argument in args:
        _utf8_size(argument, "call.args item")
    context = value["context"]
    if not isinstance(context, Mapping) or any(
        not isinstance(key, str) or not isinstance(item, str) for key, item in context.items()
    ):
        raise AdapterError("invalid_request", "call.context must be a string-to-string map")
    for key, item in context.items():
        _utf8_size(key, "call.context key")
        _utf8_size(item, "call.context value")
    return PolicyCall(
        agent=value["agent"],
        tool=value["tool"],
        args=list(args),
        user=value["user"],
        context=dict(context),
    )


def _normalize_policy_call(value: Any) -> PolicyCall:
    if not isinstance(value, PolicyCall):
        raise AdapterError("invalid_request", "call must be a PolicyCall")
    return _normalize_call(
        {
            "agent": value.agent,
            "tool": value.tool,
            "args": value.args,
            "user": value.user,
            "context": value.context,
        }
    )


def _copy_call(call: PolicyCall) -> PolicyCall:
    return PolicyCall(
        agent=call.agent,
        tool=call.tool,
        args=list(call.args),
        user=call.user,
        context=dict(call.context),
    )


def _call_dict(call: PolicyCall) -> dict[str, Any]:
    return {
        "agent": call.agent,
        "tool": call.tool,
        "args": list(call.args),
        "user": call.user,
        "context": dict(call.context),
    }


def _policy_identity(snapshot: PolicySnapshot) -> dict[str, Any]:
    return {"version": snapshot.version, "sha256": snapshot.digest}

File: src/adapter.rs
use crate::audit::AuthorizationMetadata;
use crate::models::{Action, PolicyCall};
use crate::simulator::PolicyIdentity;
use crate::AuditedPolicyStore;
use serde::Serialize;
use serde_json::{Map, Value};
use std::collections::BTreeMap;
use std::error::Error;
use std::fmt::{Display, Formatter};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

const MAX_IDENTIFIER_BYTES: usize = 256;

/// A complete, normalized request crossing the inline enforcement boundary.
#[derive(Debug, Clone)]
pub struct EnforcementRequest {
    pub caller_id: String,
    pub correlation_id: String,
    pub deadline_unix_ms: u64,
    pub call: PolicyCall,
}

impl EnforcementRequest {
    /// Parses a strict JSON-compatible request for hosts that receive request data dynamically.
    pub fn from_value(value: &Value) -> Result<Self, AdapterError> {
        let fields = value.as_object().ok_or(AdapterError::InvalidRequest)?;
        require_exact_fields(
            fields,
            &["caller_id", "correlation_id", "deadline_unix_ms", "call"],
        )?;
        let request = Self {
            caller_id: required_identifier(fields, "caller_id")?,
            correlation_id: required_identifier(fields, "correlation_id")?,
            deadline_unix_ms: fields
                .get("deadline_unix_ms")
                .and_then(Value::as_u64)
                .ok_or(AdapterError::InvalidRequest)?,
            call: parse_call(fields.get("call").expect("required call"))?,
        };
        request.validate()?;
        Ok(request)
    }

    fn validate(&self) -> Result<(), AdapterError> {
        valid_identifier(&self.caller_id)
            .then_some(())
            .ok_or(AdapterError::InvalidRequest)?;
        valid_identifier(&self.correlation_id)
            .then_some(())
            .ok_or(AdapterError::InvalidRequest)
    }
}

/// A stable failure from the enforcement boundary.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AdapterError {
    InvalidRequest,
    DeadlineExceeded,
    AuditWriteFailed,
}

impl AdapterError {
    pub const fn code(self) -> &'static str {
        match self {
            Self::InvalidRequest => "invalid_request",
            Self::DeadlineExceeded => "deadline_exceeded",
            Self::AuditWriteFailed => "audit_write_failed",
        }
    }
}

impl Display for AdapterError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(self.code())
    }
}

impl Error for AdapterError {}

/// The normalized call data returned only for an allowed policy decision.
#[derive(Debug, Clone, Serialize)]
pub struct EffectCall {
    pub agent: Option<String>,
    pub tool: Option<String>,
    pub args: Vec<String>,
    pub user: Option<String>,
    pub context: BTreeMap<String, String>,
}

impl From<&PolicyCall> for EffectCall {
    fn from(call: &PolicyCall) -> Self {
        Self {
            agent: call.agent.clone(),
            tool: call.tool.clone(),
            args: call.args.clone(),
            user: call.user.clone(),
            context: call.context.clone(),
        }
    }
}

/// Data the host may execute after a successful authorization; this type has no executor path.
#[derive(Debug, Clone, Serialize)]
pub struct EffectRequest {
    pub caller_id: String,
    pub correlation_id: String,
    pub deadline_unix_ms: u64,
    pub policy: PolicyIdentity,
    pub decision: Action,
    pub matched_rules: Vec<usize>,
    pub call: EffectCall,
}

/// An audited policy response. `effect` is absent unless `decision` is `allow`.
#[derive(Debug, Clone, Serialize)]
pub struct EnforcementResponse {
    pub caller_id: String,
    pub correlation_id: String,
    pub deadline_unix_ms: u64,
    pub policy: PolicyIdentity,
    pub decision: Action,
    pub matched_rules: Vec<usize>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub effect: Option<EffectRequest>,
}

/// Host-local enforcement boundary around an audited active policy.
pub struct InlineEnforcementAdapter {
    store: AuditedPolicyStore,
    now_unix_ms: Arc<dyn Fn() -> u64 + Send + Sync>,
}

impl InlineEnforcementAdapter {
    pub fn new(store: AuditedPolicyStore) -> Self {
        Self::with_clock(
            store,
            Arc::new(|| {
                SystemTime::now()
                    .duration_since(UNIX_EPOCH)
                    .unwrap_or_default()
                    .as_millis()
                    .try_into()
                    .unwrap_or(u64::MAX)
            }),
        )
    }

    /// Supplies a deterministic clock for tests and hosts with a monotonic wall-clock source.
    pub fn with_clock(
        store: AuditedPolicyStore,
        now_unix_ms: Arc<dyn Fn() -> u64 + Send + Sync>,
    ) -> Self {
        Self { store, now_unix_ms }
    }

    /// Validates, audits, and returns data for the host to execute only after an allow decision.
    pub fn authorize(
        &mut self,
        request: EnforcementRequest,
    ) -> Result<EnforcementResponse, AdapterError> {
        request.validate()?;
        if request.deadline_unix_ms <= (self.now_unix_ms)() {
            return Err(AdapterError::DeadlineExceeded);
        }

        let authorization = self
            .store
            .authorize_with_metadata(
                &request.call,
                Some(AuthorizationMetadata {
                    correlation_id: request.correlation_id.clone(),
                    caller_id: request.caller_id.clone(),
                    deadline_unix_ms: request.deadline_unix_ms,
                }),
            )
            .map_err(|_| AdapterError::AuditWriteFailed)?;
        let policy = PolicyIdentity {
            version: authorization.snapshot.version,
            sha256: authorization.snapshot.digest.clone(),
        };
        let effect = (authorization.decision.action == Action::Allow).then(|| EffectRequest {
            caller_id: request.caller_id.clone(),
            correlation_id: request.correlation_id.clone(),
            deadline_unix_ms: request.deadline_unix_ms,
            policy: policy.clone(),
            decision: authorization.decision.action,
            matched_rules: authorization.decision.matched_rules.clone(),
            call: EffectCall::from(&request.call),
        });

        Ok(EnforcementResponse {
            caller_id: request.caller_id,
            correlation_id: request.correlation_id,
            deadline_unix_ms: request.deadline_unix_ms,
            policy,
            decision: authorization.decision.action,
            matched_rules: authorization.decision.matched_rules,
            effect,
        })
    }
}

fn valid_identifier(value: &str) -> bool {
    !value.is_empty() && value.len() <= MAX_IDENTIFIER_BYTES
}

fn require_exact_fields(
    fields: &Map<String, Value>,
    required: &[&str],
) -> Result<(), AdapterError> {
    if required.iter().any(|field| !fields.contains_key(*field))
        || fields
            .keys()
            .any(|field| !required.contains(&field.as_str()))
    {
        return Err(AdapterError::InvalidRequest);
    }
    Ok(())
}

fn required_identifier(fields: &Map<String, Value>, field: &str) -> Result<String, AdapterError> {
    let value = fields
        .get(field)
        .and_then(Value::as_str)
        .filter(|value| valid_identifier(value))
        .ok_or(AdapterError::InvalidRequest)?;
    Ok(value.to_owned())
}

fn parse_call(value: &Value) -> Result<PolicyCall, AdapterError> {
    let fields = value.as_object().ok_or(AdapterError::InvalidRequest)?;
    require_exact_fields(fields, &["agent", "tool", "args", "user", "context"])?;
    let nullable_string = |field: &str| match fields.get(field) {
        Some(Value::Null) => Ok(None),
        Some(Value::String(value)) => Ok(Some(value.clone())),
        _ => Err(AdapterError::InvalidRequest),
    };
    let args = fields
        .get("args")
        .and_then(Value::as_array)
        .and_then(|values| {
            values
                .iter()
                .map(|value| value.as_str().map(str::to_owned))
                .collect::<Option<Vec<_>>>()
        })
        .ok_or(AdapterError::InvalidRequest)?;
    let context = fields
        .get("context")
        .and_then(Value::as_object)
        .and_then(|values| {
            values
                .iter()
                .map(|(key, value)| value.as_str().map(|value| (key.clone(), value.to_owned())))
                .collect::<Option<BTreeMap<_, _>>>()
        })
        .ok_or(AdapterError::InvalidRequest)?;
    Ok(PolicyCall {
        agent: nullable_string("agent")?,
        tool: nullable_string("tool")?,
        args,
        user: nullable_string("user")?,
        context,
    })
}
